### ISSUE TITLE
Two incompatibilities with ds and dS defined from ufl.Measure

### ISSUE THREAD

**Issue by francesco-ballarin** [FEniCS/dolfinx#536]
Dear developers,
I am facing two issues when using ds and dS measures.

I have customized test_ghost_mesh_assembly.py in
https://github.com/francesco-ballarin/dolfinx/commit/63b57bac2
to add a test parametrization on dx as follows:
  1. `dx_from_ufl`: use ufl.dx, as in master
  2. `dx_from_measure`: use ufl.Measure("dx", subdomain_data=marker), where marker is the constant 1 on the whole mesh
   3. `dx_from_measure_and_subdomain`: take the output of `dx_from_measure` and apply a restriction to the subdomain 1, which effectively should be the whole domain

and similarly for ds and dS.

As soon as `ds_from_measure_and_subdomain` is used, tests are failing
```
test_ghost_mesh_assembly.py::test_ghost_mesh_assembly[ds_from_measure_and_subdomain-dx_from_ufl-mode0] FAILED
test_ghost_mesh_assembly.py::test_ghost_mesh_assembly[ds_from_measure_and_subdomain-dx_from_measure-mode0] FAILED
test_ghost_mesh_assembly.py::test_ghost_mesh_assembly[ds_from_measure_and_subdomain-dx_from_measure_and_subdomain-mode0] FAILED
```
and an incorrect norm is obtained
```
>       assert normA == pytest.approx(0.6713621455570528, rel=1.e-6, abs=1.e-12)
E       assert 2.7208495255521115 == 0.6713621455570528 ± 6.7e-07
E         -2.7208495255521115
E         +0.6713621455570528 ± 6.7e-07
```
I believe this to be an incompatibility between sums of integrals in ds and dx, because `test_assembly_ds_domains` (which only has integrals in ds) is working.

Furthermore, the test with dS(1)
```
test_ghost_mesh_assembly.py::test_ghost_mesh_dS_assembly[dS_from_measure_and_subdomain-mode0] FAILED
```
is failing with
```
    @functools.singledispatch
    def assemble_matrix(a,
                        bcs: typing.List[DirichletBC] = [],
                        diagonal: float = 1.0) -> PETSc.Mat:
        """Assemble bilinear form into a matrix. The returned matrix is not
        finalised, i.e. ghost values are not accumulated.
    
        """
        a_cpp = _create_cpp_form(a)
        A = cpp.fem.create_matrix(a_cpp)
        A.zeroEntries()
>       cpp.fem.assemble_matrix(A, a_cpp, bcs, diagonal)
E       RuntimeError: Mesh entity was not found
```

**Comment by chrisrichardson** [comment]
I have a feeling that it may be reading the `MeshFunction` a bit too much.
 
For example, I suppose for the `ds` integrals, if you set the subdomain to be "1" on every facet, and then use `ds(1)` it will integrate on every facet, (including any interior facets) and not just on the surface facets. Likewise with the `dS(1)` integrals, it will try to integrate them on exterior facets (if you set the `MeshFunction` to 1 on those facets).  

Perhaps this is useful sometimes (for `ds`)?

But really it is also wasteful (especially on a big mesh) to set a `MeshFunction` for every facet, when we are only considering the surface... I'm not sure what is the best solution. Maybe you can confirm, but setting the `MeshFunction` only for interior or exterior facets, and see if that helps. Then we can think about a better design.

**Comment by francesco-ballarin** [comment]
Thanks for the comments. I pushed the updated version at
https://github.com/francesco-ballarin/dolfinx/commit/0c4259bbf96787dec4c2564a6ab467e7a4f45c2c

This does indeed fix the failing tests in serial. However, failures still occurs in parallel (wrong norm for ds tests, RuntimeError for dS).

> Perhaps this is useful sometimes (for ds)?

Apologies, I don't follow. Then what would be the point of dS?

> But really it is also wasteful (especially on a big mesh) to set a MeshFunction for every facet

Apologies again. Isn't the MeshFunction value set on all facets at construction time? What would be wasteful in setting it to one rather than zero?

To give a concrete example, I sometimes use the same MeshFunction (which is read from file, and nonzero on boundaries and some internal interfaces) to setup both ds and dS. Of course that can easily change, just to let you know how is it that I first encountered this issue.

**Comment by chrisrichardson** [comment]
Internally, there has been a change, so that the assembler now accepts a list of "active_facets", rather than a `MeshFunction`. I suppose this is more like a `MeshValueCollection`, conceptually. Of course, there are advantages and disadvantages to both approaches.

I guess all we need to do is filter the `MeshFunction` appropriately, so that it only applies to the internal (for `dS`) and external (for `ds`) facets. At the moment, that filtering is not being done.
When I said it was wasteful, I meant that if you are just interested in `ds`, then there are a lot of wasted facets.

As for the usefulness of `ds` integrals on internal facets - I did say "perhaps". I *can* see a use for it (since `dS` always uses two cells, whereas `ds` is only on one)... but it is a very limited use case.

**Comment by francesco-ballarin** [comment]
OK, thanks for the clarification. I am available to test the implementation of the filter on MeshFunction, when it will be ready. In the meantime, let me know if you want me to make a pull request of the updated test.

**Comment by chrisrichardson** [comment]
I have made a change in the branch `chris/restrict-facet-integral-markers`, if you are able to try it.

**Comment by francesco-ballarin** [comment]
Thanks, I rebased and pushed to
https://github.com/francesco-ballarin/dolfinx/tree/ds_dS_errors
using the original test in the opening post (not the one in the second post where the restriction was manually applied). I also added a couple of minor fixes.

I am still getting failures as in the original post (incorrect values are computed in all failing tests) 
```
test_ghost_mesh_assembly.py::test_ghost_mesh_assembly[ds_from_measure_and_subdomain-dx_from_ufl-mode0] FAILED
test_ghost_mesh_assembly.py::test_ghost_mesh_assembly[ds_from_measure_and_subdomain-dx_from_measure-mode0] FAILED
test_ghost_mesh_assembly.py::test_ghost_mesh_assembly[ds_from_measure_and_subdomain-dx_from_measure_and_subdomain-mode0] FAILED
test_ghost_mesh_assembly.py::test_ghost_mesh_dS_assembly[dS_from_measure_and_subdomain-mode0] 0.062112999374994156
FAILED
````

**Comment by chrisrichardson** [comment]
It would be nice to break down a bit the tests, so we can see what is actually causing the failure. I think it may now be the `dx` integrals for the ghosted mesh, maybe assembling ghost cells twice.

**Comment by chrisrichardson** [comment]
I think it might now be OK for `dx` and `ds`. I am not so confident about `dS` as there are some serious problems with "ownership" of facets on ghosted meshes. There is another issue #506 which should eventually deal with it. I'll try and hack around it if I can for now.

**Comment by francesco-ballarin** [comment]
Thanks, I confirm that all tests are now passing both in serial and in parallel. Feel free to cherrypick my commit https://github.com/francesco-ballarin/dolfinx/commit/13a07be4603e49d4da1a6368cd8e899c71620e14 in your branch if you want to include the tests.

Please allow me to have a couple more days before closing the issue. I might have a further related problem, in which assembling with `assemble_scalar` two scalar forms, one in ds and the other one in dS separately, and then summing the scalar results, gives the correct answer, but summing the forms and then assemblying gives a wrong answer. I need a couple more days to understand if it is an error in my code, or if I can reproduce slightly changing one of the existing tests on master.

**Comment by francesco-ballarin** [comment]
Hi,
please find at https://github.com/francesco-ballarin/dolfinx/commit/2bb0ca88f7fa10c693c56eaa74a52dddd1232d9f the test related to my previous comment. They all fail as soon as dS is used in the sum.

**Comment by chrisrichardson** [comment]
This is an unrelated issue, but needs fixing. I'll start a new issue.

**Comment by francesco-ballarin** [comment]
This was fixed by #543, the remaining issue is being discussed in #545. Closing.
